minio/minio
# B2 gateway S3 compat: return MD5 hash as ETag from PutObject

## Description

Quoting the B2 Gateway Readme:

> PutObject S3 API does not return the md5sum of an uploaded file as it's etag response. Applications that check the validity of the uploaded object by comparing these two values will fail.

However, as confirmed by the docs, and as tested in both US and EU regions, B2 actually does return an MD5 hash for a stored file via the https://www.backblaze.com/b2/docs/b2_get_file_info.html endpoint.

This we can use to provide better compatibility with S3 client libraries that assume the ETag is the MD5 hash - such as boto.

The change depends on a PR for blazer: https://github.com/kurin/blazer/pull/70

I opted to only enable this change if MinIO's `--compat` mode is active, however this can be discussed.

## Motivation and Context

I was unable to use MinIO for migrating about 700.000 files from local storage to B2 because of the error described below. This is a Seafile server, which uses the [boto library](https://github.com/boto/boto) in version 2.49, however the MD5-ETag requirement also seems to exist in boto3: https://github.com/boto/boto3/issues/544

**This PR is not ready to merge - I would like to get some feedback on whether this approach seems acceptable to you. Definitely missing are: unit tests, updated go.mod for blazer library.**

## How to test this PR?

- Run MinIO in B2 Gateway mode. Try to upload a file via a tool that uses boto. Run into the following error:
`BotoClientError: ETag from S3 did not match computed MD5. "4_...-1" vs. b29c551c2decea1c0625b34fda00db09
`
- This usually causes the file to exist in the B2 bucket because most clients don't remove it after running into this error.
- Apply patches and start MinIO with `--compat`.
- Re-try upload, should succeed without an error.

## Types of changes

## Checklist:
- [x] Discussion of approach with `--compat` flag
- [x] Updated go.mod with changes for blazer library
